common/src/main/java/com/wynntils/features/map/WorldMarkersFeature.java

```diff
@@ -0,0 +1,539 @@
+        g = MathUtils.clamp(g, 0, 255);
+        b = MathUtils.clamp(b, 0, 255);
+        currentRainbowColor = new CustomColor(r, g, b, 255);
+    }
+
+    // Rendered map location calculation happens here
+    // Beacon beam rendering happens here
+    @SubscribeEvent
+    public void onRenderLevelLast(RenderTileLevelLastEvent event) {
+        // We update the rendered map locations here so that we can render the beacon beams in the same event
+        // and also use it below to render the icons and labels
+        updateRenderedMapLocations(event.getProjectionMatrix(), event.getCamera());
+        if (renderedMapLocations.isEmpty()) return;
+
+        PoseStack poseStack = event.getPoseStack();
+
+        // render beacon beams
+        for (RenderedMapLocation mapLocationPair : renderedMapLocations) {
+            double alpha = mapLocationPair.visibility();
+            MapLocation mapLocation = mapLocationPair.mapLocation();
+
+            ResolvedMapAttributes resolvedMapAttributes = Services.MapData.resolveMapAttributes(mapLocation);
+            ResolvedMarkerOptions resolvedMarkerOptions = resolvedMapAttributes.markerOptions();
+
+            Position camera = event.getCamera().getPosition();
+            Location location = mapLocation.getLocation();
+
+            double dx = location.x - camera.x();
+            double dy = location.y - camera.y();
+            double dz = location.z - camera.z();
+
+            double distance = MathUtils.magnitude(dx, dz);
+            int maxDistance = McUtils.options().renderDistance().get() * 16;
+
+            if (distance > maxDistance) {
+                double scale = maxDistance / distance;
+
+                dx *= scale;
+                dz *= scale;
+            }
+
+            poseStack.pushPose();
+            poseStack.translate(dx, dy, dz);
+
+            CustomColor color = resolvedMarkerOptions.beaconColor() == CustomColor.NONE
+                    ? markerBeaconColor.get()
+                    : resolvedMarkerOptions.beaconColor();
+
+            int colorInt;
+            if (color == CommonColors.RAINBOW) {
+                colorInt = currentRainbowColor.withAlpha((float) alpha).asInt();
+            } else {
+                colorInt = color.withAlpha((float) alpha).asInt();
+            }
+
+            BeaconRenderer.renderBeaconBeam(
+                    poseStack,
+                    BUFFER_SOURCE,
+                    BeaconRenderer.BEAM_LOCATION,
+                    event.getDeltaTracker().getGameTimeDeltaPartialTick(false),
+                    1f,
+                    McUtils.player().level().getGameTime(),
+                    0,
+                    BeaconRenderer.MAX_RENDER_Y,
+                    colorInt,
+                    0.166f,
+                    0.33f);
+
+            poseStack.popPose();
+        }
+
+        BUFFER_SOURCE.endBatch();
+    }
+
+    private void updateRenderedMapLocations(Matrix4f projectionMatrix, Camera camera) {
```

Comment:
This is done one for every rendered frame? My gut feeling is that this is way too heavy for that.

Reply:
It's basically has been like this since the original feature existed, it's not too much calculations, if anything, it's collecting the map locations that might introduce some overhead, but since basically all of the attributes get cached, it's basically just a glorified getter. Before merging the overall mapdata branch, I do plan to profile map and marker performance, and if anything comes up then, we can fix it.